### PR TITLE
Allocated Classes

### LINKED ISSUE #161: Provide mechanism for creating new boxed version of a COM class casted as a particular interface.
I'm playing around with the latest revision of this crate (currently 3693ab2e1); I'm trying to implement some interfaces for a custom type, and I'm trying to figure out how to get one of the interface I implement once I have created an instance my concrete 'class'.

afaict, the only way I can do this is by using `query_interface`, and doing a bunch of pointer casts. I see that there's a `get_interface` method defined on the interface type, but it isn't defined on my class, so it doesn't seem like I can use it? The best I can do so far in terms of some kind of general solution is to use a macro like,

```rust
macro_rules! get_interface {
    ($item:expr, $interface:ty) => {{
        let mut ppv: Option<$interface> = None;
        let iid = &<$interface as com::Interface>::IID as *const _;
        let hr = unsafe {
            $item.query_interface(iid, &mut ppv as *mut _ as *mut *mut c_void)
        };
        if SUCCEEDED(hr) {
            ppv
        } else {
            None
        }
    }}
}
```

and then I can call this like, `let interface = get_interface(my_type, IDWriteFontFile)`, but it doesn't feel like a great solution, but since query_interface isn't part of a trait I don't see an alternative?

I'm probably missing something obvious, but any pointers here would be appreciated. :)

### PR DESCRIPTION
**Pull request by rylev**
Fixes #161 

This (hopefully) fixes the issues in #161 which are:
* **COM interface methods must be called on objects that are heap allocated**: We don't force the user to conform to this for regular methods because it's possible to safely use methods with `&self`. However, the IUnknown methods assume that self is heap allocated and won't move. We model this by taking self as `self: &Pin<Box<Self>>`. This makes it not possible to call IUnknown methods on a stack allocated object. If users must assume that their code is heap allocated they can declare their methods with the same signature as the IUnknown methods but if the assumption is not needed, they can just use `&self`. 
* **There is no safe query_interface**: classes now come with a safe query interface method called `query`. This may only be used when the type is heap allocated and pinned. 
* **It is confusing how to properly instantiate a class**:  Classes now come with an `alloc` associated function which heap allocates the class, pins it, and queries the class for a given interface. 
* **Memory management of the class is potentially wrong**: classes now properly use their ref count to manage memory. The ref count starts at 1 and is incremented and decremented with `add_ref` and `release` respectively. On drop, the ref count is also decremented. When the count is 0, the class's contents are dropped. This should only happen inside of the `release` thunk or if the class is manually allocated by the user and never shared.

### CODE REVIEW

File: macros/support/src/class/iunknown_impl.rs [hunk 1]
```diff
@@ -35,25 +35,28 @@ impl IUnknownAbi {
         quote! {
             unsafe extern "stdcall" fn release(this: #this_ptr) -> u32 {
                 #munge
-                #class_name::release(&this)
+                let count = #class_name::release(&munged);
+                if count == 1 {
```
Comment (rylev) [inline macros/support/src/class/iunknown_impl.rs:39]: I'm not sure how I feel about this. We're relying on a ref count of 1 meaning that the only handle to the class is the original one which should have been "forgotten" through the use of `ManuallyDrop`. If the user forgets to wrap the class in a `ManuallyDrop`, and calls `query` to get an interface then it will have a ref count of 1 (the handle to the interface requested in `query`. When `release` is called on that interface the count will be 0. This is unfortunate as it leaks memory, though it is not unsafe. 

An alternate implementation would drop the class when the count is 0. However, normally, this would leak memory in the normal case of "forgetting" the class when created by a class factory.

Comment (rylev) [reply macros/support/src/class/iunknown_impl.rs:39]: It's actually also possible to have a dangling pointers if this function is called with a class that has not been "forgotten":

If I return an interface from IClassFactory::create_instance that has been created from a class struct that was then dropped, it's ref count will be 1 (just the handle to interface instead of 2 for the handle and the original class object). If I then make another handle, and it gets released through the call to release ABI function, it will free the memory even though there is still a handle to an interface out there. 

This requires improperly implementing an unsafe function though so I don't think it's not an obviously wrong way to go.

Comment (kennykerr) [reply macros/support/src/class/iunknown_impl.rs:39]: The usual way to handle this is for the "make" function to allocate the object and return the only reference to the object as a COM interface. That COM interface holds that first reference count.

Comment (rylev) [reply macros/support/src/class/iunknown_impl.rs:39]: Yes, this is what we're currently doing, though the user can allocate on their own (i.e., there's nothing stopping them other than us documenting that they shouldn't). Our "make" function is called `alloc` and it does just this. I'm looking for a way to prevent users from holding on to COM classes at all since they should use `alloc`.

Comment (kennykerr) [reply macros/support/src/class/iunknown_impl.rs:39]: Ah, so in C++/WinRT we have a virtual function called [use_make_function_to_create_this_object](https://kennykerr.ca/2019/05/08/cppwinrt-diagnosing-direct-allocations/) that solves this problem. Perhaps in Rust we could have the extra fields needed for COM (e.g. vfptrs and ref count) such that they are not `Default` so you cannot simply construct the object yourself.

Comment (rylev) [reply macros/support/src/class/iunknown_impl.rs:39]: Yea I'm currently thinking how we want to do this. The code is being generated by the macro so it's currently as if the user wrote the code themselves (i.e., everything the macro produces, the user must also be able to write themselves). I'm thinking about how we can give the macro the ability to create the class but not give this ability to the user.

Comment (rylev) [reply macros/support/src/class/iunknown_impl.rs:39]: I pushed a change that should help. 

It's still technically possible inside of the module where the class is declared to instantiate the class on the stack, but this has be done by hand and is not trivial (it's about 225 lines of total code in the basic server example). The class is not Default so there's no way to instantiate that way (though you can call `MyClass::allocate_default()` to get a proper heap allocated version where all the user fields are defaulted). Outside of the current module it's not possible at all to stack allocate the class. 

Obviously we need docs to clarify how this all should work, but it's now highly unlikely that the user will do the wrong thing.

File: examples/basic/client/src/main.rs [hunk 2]
```diff
@@ -6,56 +6,73 @@ use com::{
...
     init_apartment(ApartmentType::SingleThreaded)
         .unwrap_or_else(|hr| panic!("Failed to initialize COM Library{:x}", hr));
+    println!("Initialized apartment");
 
+    // Get a `BritishShortHairCat` class factory
     let factory = get_class_object::<IClassFactory>(&CLSID_CAT_CLASS)
         .unwrap_or_else(|hr| panic!("Failed to get cat class object 0x{:x}", hr));
     println!("Got cat class object");
 
+    // Get an instance of a `BritishShortHairCat` as the `IUnknown` interface
```
Comment (kennykerr) [inline examples/basic/client/src/main.rs:19]: I assume this is just an example and not a requirement. 😉

File: examples/basic/client/src/main.rs [hunk 3]
```diff
@@ -6,56 +6,73 @@ use com::{
 use interface::{Food, IAnimal, ICat, IDomesticAnimal, IExample, CLSID_CAT_CLASS};
 
 fn main() {
+    // Initialize the COM apartment
     init_apartment(ApartmentType::SingleThreaded)
         .unwrap_or_else(|hr| panic!("Failed to initialize COM Library{:x}", hr));
+    println!("Initialized apartment");
 
+    // Get a `BritishShortHairCat` class factory
     let factory = get_class_object::<IClassFactory>(&CLSID_CAT_CLASS)
```
Comment (kennykerr) [inline examples/basic/client/src/main.rs:15]: Is there also an equivalent of `CoCreateInstance(Ex)`?

File: examples/basic/client/src/main.rs [hunk 4]
```diff
@@ -6,56 +6,73 @@ use com::{
...
         .unwrap_or_else(|hr| panic!("Failed to get cat class object 0x{:x}", hr));
     println!("Got cat class object");
 
+    // Get an instance of a `BritishShortHairCat` as the `IUnknown` interface
     let unknown = factory
         .get_instance::<IUnknown>()
```
Comment (kennykerr) [inline examples/basic/client/src/main.rs:21]: Why is `CreateInstance` renamed to `get_instance`?

File: examples/basic/client/src/main.rs [hunk 5]
```diff
@@ -6,56 +6,73 @@ use com::{
...
         .expect("Failed to get IUnknown");
     println!("Got IUnknown");
 
+    // Now get a handle to the `IAnimal` interface
     let animal = unknown
         .get_interface::<IAnimal>()
```
Comment (kennykerr) [inline examples/basic/client/src/main.rs:27]: Why is `QueryInterface` renamed to `get_interface`?

File: macros/support/src/class/class_constructor.rs [hunk 6]
```diff
@@ -2,48 +2,111 @@ use super::{class::Interface, Class};
...
+            let instance = ::std::mem::ManuallyDrop::new(::std::boxed::Box::pin(instance));
+            instance.query()
         }
     }
 }
 
-// User field input as parameters to the allocate function.
-fn gen_allocate_user_fields(class: &Class) -> TokenStream {
-    let field_idents = class.fields.iter().map(|f| &f.ident);
+/// Function used to instantiate a default version COM class
```
Comment (kennykerr) [inline macros/support/src/class/class_constructor.rs:39]: Not sure what this comment is trying to tell me.

File: examples/basic/client/src/main.rs (hunk 2 above)
Comment (rylev) [reply examples/basic/client/src/main.rs:19]: Do you mean you hope it's not a requirement to get the IUnknown interface? It's not as `get_instance` is generic over the interface you want to get the class as.

File: examples/basic/client/src/main.rs (hunk 3 above)
Comment (rylev) [reply examples/basic/client/src/main.rs:15]: This is a call to CoGetClassObject. We use `create_instance` below which directly maps to `CoCreateInstance`.

File: examples/basic/client/src/main.rs (hunk 4 above)
Comment (rylev) [reply examples/basic/client/src/main.rs:21]: This is because there is no method overloading. I decided to rename the safe functions since they have different signatures and thus calling the safe version `create_instance` might be confusing. I'm starting to think that for the raw win32 COM API calls we should use the CamelCase names and the safe versions can use the snake_case convention. This might be a better way to distinguish the two. Thoughts?

File: examples/basic/client/src/main.rs (hunk 5 above)
Comment (rylev) [reply examples/basic/client/src/main.rs:27]: Same as above. This is the safe wrapper around `query_interface`. We could rename the unsafe version `QueryInterface` and the safe version could then be `query_interface`.

File: macros/support/src/class/class_constructor.rs (hunk 6 above)
Comment (rylev) [reply macros/support/src/class/class_constructor.rs:39]: Whoops. This is a function used to create an instance of the class where all the user fields are default. We don't want to provide a normal `Default` implementation because that would allow the user to stack allocate the class which we don't want.

File: macros/support/src/class/class.rs [hunk 7]
```diff
@@ -123,17 +123,39 @@ impl Class {
...
-        let iunknown = super::iunknown_impl::IUnknown::new(name.clone());
+        let iunknown = super::iunknown_impl::IUnknown::new();
         let add_ref = iunknown.to_add_ref_tokens();
-        let release = iunknown.to_release_tokens(interfaces);
+        let release = iunknown.to_release_tokens();
         let query_interface = iunknown.to_query_interface_tokens(interfaces);
+        let query = iunknown.to_query_tokens();
         let constructor = super::class_constructor::generate(self);
+        let default_constructor = super::class_constructor::generate_default(self);
+        let unsafe_constructor = super::class_constructor::generate_unsafe(self);
```
Comment (MarijnS95) [inline macros/support/src/class/class.rs:143]: Similar to `derive(Debug)` a while ago not all user-specified members may have `Default` implemented; this currently makes it impossible to create a COM class with such members.

Would it be possible to emit these only when the `class` definition is annotated with `#[derive(Default)]`?

Comment (rylev) [reply macros/support/src/class/class.rs:143]: I think we don't want to allow annotating the class with `#[derive(Default)]` as either this creates a `Default::default` implementation which would allow for stack allocation (which we don't want to allow) or it doesn't which is confusing. I think we'll need some other attribute to indicate that your type is defaultable. This is needed for use with `CoCreateInstance` as the implementation of that function creates an instance of the class. I'll need to think more about this.

Comment (rylev) [reply macros/support/src/class/class.rs]: @MarijnS95 I added the ability to add a `#[no_class_factory]` attribute to your class. This removes the class factory generation (meaning you can't use `CoCreateInstance`) along with any need for your data members to be Default

**Comment by rylev** [comment]
@kennykerr - I think this PR is ready to go. You had some concerns about naming, but that's not really a part of this PR (as that exists in master). Shall we merge this and open an issue for your concerns about names?

File: macros/support/src/class/class.rs (hunk 7 above)
Comment (MarijnS95) [reply macros/support/src/class/class.rs]: I'll probably chain that into Linux support at some point as `CoCreateInstance` isn't available there anyway.

`generate_unsafe` is still there emitting an almost identical constructor to `generate_default` including forced defaults for members. I guess that shouid be behind `has_class_factory` as well, which is what you need this for?

Besides `allocate` has to return some type that implements the `Interface` trait, which is not the case for the `class` itself. This means it's currently impossible to receive `Self` (or a `Pin<Box<Self>>`) to interact with these members after constructing. Is that intended?

Comment (rylev) [reply macros/support/src/class/class.rs]: Oh whoops. Forgot to not return early from `generate_unsafe` if `has_class_factory` is false. Will change that.

Yes this is on purpose but that's usually what you want. Can you explain your use case? You want to be able to call a regular method on the class object without going through an interface?

Comment (MarijnS95) [reply macros/support/src/class/class.rs]: > Yes this is on purpose but that's usually what you want. Can you explain your use case? You want to be able to call a regular method on the class object without going through an interface?

We don't have such a usecase yet, but it doesn't seem uncommon to access functions and fields of the class that are not exposed through any COM interface, because they are only supposed to be touched (directly) from the hosting application.

Comment (rylev) [reply macros/support/src/class/class.rs]: Hmmm... this is a tough one which is why I'd like to see a use case first. Again, the class implementation _must_ live for as long as there's are interface handles to that class. If we were to have a function that returns a `Pin<Box<MyClasss>>`, the user would have to remember to ensure it's not dropped if any interface handles are given out. We _could_ have a function that returns a `ManuallyDrop<Pin<Box<MyClass>>>`, but that's something I'd like to add first when we have a use case for it.

Comment (MarijnS95) [reply macros/support/src/class/class.rs]: > the user would have to remember to ensure it's not dropped if any interface handles are given out

That's pretty much the point I tried to get across in https://github.com/microsoft/com-rs/issues/161#issuecomment-674873304; internal refcounting and free-semantics should be oblivious to the user.

Can the function return a reference to `MyClass`? It's after all held in-place by `ManuallyDrop<Pin<Box<MyClass>>>` internally.

Comment (rylev) [reply macros/support/src/class/class.rs]: COM interfaces are directly analogous to Rust's [Rc](https://doc.rust-lang.org/stable/std/rc/struct.Rc.html) not to references.

A COM class makes no sense unless it ultimately becomes a set of interface handles. After all, if you don't hand off interfaces to other code, why did you make this a COM class in the first place? We could allocate a COM class and then hand a reference to it, but then we have no interface handles, making the class pointless. 

Again, if you could give me an example use case, it would be much easier to discuss this.

Comment (MarijnS95) [reply macros/support/src/class/class.rs]: > COM interfaces are directly analogous to Rust's [Rc](https://doc.rust-lang.org/stable/std/rc/struct.Rc.html) not to references.

In that sense the class object should be _moved into_ the `Rc`; only accessing it through the `Rc` afterwards.

> After all, if you don't hand off interfaces to other code

Not exposing a certain member through a COM interface doesn't mean I might not want to access it at all anymore from Rust. Right now that likely has to be wrapped in `(A)Rc` before a class instance is constructed so that it can be accessed by both parties.

> why did you make this a COM class in the first place?

To _connect_ Rust code to COM interfaces.

> We could allocate a COM class and then hand a reference to it, but then we have no interface handles, making the class pointless.

And we can't have both? A class that contains some local (to the Rust program) fields, exposing a COM interface that administers interop between the fields and functions in the interface?

> Again, if you could give me an example use case, it would be much easier to discuss this.

I could probably write one up though it would be artificial; consider `fn happiness(&self)` lacking from `IAnimal` because happiness is something local to our Rust program. If I want to access the `happiness` field after interacting with some `COM` classes I have the following ways of interacting with it:

- Creating a `happiness` variable in a higher scope, referenced through a `RefCell` within the class;
- Create a (local, private) COM interface just to access it;
- [What I'm asking for] have ~direct access to the class `struct`/`impl.

Comment (rylev) [reply macros/support/src/class/class.rs]: @MarijnS95, ok I made a change which I think addresses this. `allocate` now returns a `Pin<Box<MyClass>>` with the ref count set to 1. If you call `query` the ref count naturally goes to 2. Dropping the class, causes the ref count to go back to one. In `release` (which is called when dropping the interface you got from calling `query`) puts the ref count at 0 and drops the associated resources like the vtables and the user defined fields. 

We should probably add a `Clone` implementation to each class so that you can have more direct handles to the class (by incrementing the ref count).

Comment (MarijnS95) [reply macros/support/src/class/class.rs]: Cool!

Unsurprisingly it segfaults because the `release` function in `IUnknownAbi` always drops regardless of (decremented) refcount: https://github.com/microsoft/com-rs/pull/163/commits/f079c350128a414747a22d6134cd20ba45cc51b7#diff-bd234ac685dab7ad1d70a35db06cc905L38-R39.

Then I have to make sure to keep the `Pin<Box<MyClass>>` around (and use references to it) rather than moving it to prevent it being dropped. Adding logs to all explicit drops I managed to find in `com-rs` show it's not any of those. For reference: https://github.com/Traverse-Research/hassle-rs/commit/692a7a4da91aeade62412772f33f2c27b04efe6b#diff-75009d71fc672c2f26529de8c1281c18R266-R269 if I uncomment those the segfault (that appears because the vtable is accessed after being freed) disappears.

Comment (rylev) [reply macros/support/src/class/class.rs]: @MarijnS95  ok then we defintiely still have a bug. You should be able to drop the main class after you've called `query` and not have anything crash. `Pin<Box<MyClass>>` means the class is heap allocated (that's the `Box` part), and you can't move it (that's the `Pin` part). 

It's true that the class gets dropped each time but dropping the class simply decrements the ref count. Only when the refcount becomes 0 does the class actually release its resources. Can you trace what the ref count is at various points in the program and see what's happening?

Comment (MarijnS95) [reply macros/support/src/class/class.rs]: @rylev in all drop/release cases I found and annotated with logs the ref count was still above 0 and no manual dropping occurred, meaning there's an uncaught case hidden somewhere.

I tried setting up a simple example mimicking an inproc dll earlier (with another class) but failed; perhaps I have to try again and use that as a more easy-to-use example and perhaps `#[test]`? That should allow easier debugging and most notably quicker validation during development.

Comment (rylev) [reply macros/support/src/class/class.rs]: > The point being that the class should not be dropped because I still have an interface to it open, and the refcount > 0

Let me rephrase this and see if you still agree. "The class should not free its resources until the refcount is 0". The difference is that a class, just like an interface and a `Rc` are really just reference counted handles. So we might call the drop impl on a class, but as long as the ref count is above 0, we won't free any of the associated resources.

With that being said, there is definitely a bug. On my machine when dropping the class handle (by dropping the class on the line you indicated), something turns the vptr to a null pointer. In other words, before the class drop impl is called, the `handler_wrapper` is a `**VTable` where both pointers are valid. After the drop impl is run the second pointer (i.e., the pointer being pointed to by `handler_wrapper` has been turned into a null pointer. The class drop impl does not do this, and I'm not sure what could be doing it. I'll keep looking into it more...

Comment (MarijnS95) [reply macros/support/src/class/class.rs]: > Let me rephrase this and see if you still agree. "The class should not free its resources until the refcount is 0". The difference is that a class, just like an interface and a `Rc` are really just reference counted handles. So we might call the drop impl on a class, but as long as the ref count is above 0, we won't free any of the associated resources.

Agreed, with the intricate detail that the resource here is the class itself, containing - among which - the refcount. It is indeed fine to ~be dropped~ call drop on this class as long as that path ends up in com-rs to apply refcounting rules before dropping the members and freeing the surrounding box allocation.

> With that being said, there is definitely a bug. On my machine when dropping the class handle (by dropping the class on the line you indicated), something turns the vptr to a null pointer. In other words, before the class drop impl is called, the `handler_wrapper` is a `**VTable` where both pointers are valid. After the drop impl is run the second pointer (i.e., the pointer being pointed to by `handler_wrapper` has been turned into a null pointer. The class drop impl does not do this, and I'm not sure what could be doing it. I'll keep looking into it more...

Yeah I added logging to all drops and releases but couldn't find any being called and destroying the class because `refcount == 0`, seems like something is yet unhandled. Good luck figuring it out :crossed_fingers: 

Meanwhile I should resurrect the inproc example and turn it into a `#[test]`, that should make validating these scenarios easier going forward.

File: examples/basic/client/src/main.rs (hunk 4 above)
Comment (kennykerr) [reply examples/basic/client/src/main.rs:21]: Yeah, I don't renaming is such a good idea. It implies they do something else.

File: examples/basic/client/src/main.rs (hunk 2 above)
Comment (kennykerr) [reply examples/basic/client/src/main.rs:19]: Good, the example just made it look like I needed to retrieve the interface and then later query for the desired interface.

File: examples/basic/client/src/main.rs (hunk 5 above)
Comment (kennykerr) [reply examples/basic/client/src/main.rs:27]: Yes that may be better.

File: macros/support/src/class/class.rs (hunk 7 above)
Comment (rylev) [reply macros/support/src/class/class.rs]: @MarijnS95 any help you can provide figuring this out would be greatly appreciated. I believe something is completely overwriting the class in memory. Even when using `as_ref` to prevent the class being dropped, the class's ref count seems to be overwritten with a junk value. It just so happens that this doesn't cause a crash.

Comment (rylev) [reply macros/support/src/class/class.rs]: So I've pinned down an issue although I'm not 100% sure it's _the_ issue, and I'm not sure what's actually the root cause. In the drop implementation of interfaces, we call call `release` on the inner pointer. For the `IDxcIncludeHandler` interface on the `DxcIncludeHandlerWrapper` class, the call to release seems to zero out the `IUnknown` interface pointer and mangle the ref count (but seems to leave the rest of the class in tact). This is despite all these fields being wrapped in `ManuallyDrop`. This does not seem to happen with other interfaces, and I can't seem to reproduce this using another class. Still need to dig more

Comment (rylev) [reply macros/support/src/class/class.rs]: I did some more sleuthing with gdb and realized that the answer is pretty obvious. If we drop a `Pin<Box<MyClass>>` it doesn't matter if the contents of the class are wrapped in `ManuallyDrop` the memory where `MyClass` is located will be freed. We've been trying to retrofit `Box<Pin<T>>` on to what is needed for a heap allocated class, but those types don't quite model the problem right. We need to return a custom smart pointer. I'll get this fixed.

Comment (MarijnS95) [reply macros/support/src/class/class.rs]: Neat, I already anticipated `Box<Pin<T>>` to do something finicky without calling any of the `Drop` implementations of `MyClass`, for some odd reason.
It indeed looks like we need to go back to a `ComPtr` wrapping the objects in a safe and meaningful manner, which is sad. Can't you simply wrap the `Pin<Box<T>>` in something that is opaque to the end user yet derefs to the underlying `MyClass` while at the same time catching the `Drop`? That might make you get away without smart pointer as this issue seems to only be applicable to this specific wrapping of a class.

Comment (rylev) [reply macros/support/src/class/class.rs]: I don't plan on switching back to `ComPtr`. I'll most likely just create a `ClassObject<T>` wrapper object that does the right thing.

Comment (rylev) [reply macros/support/src/class/class.rs]: @MarijnS95 I pushed some new code, and it seems to fix the segfault issue in hassle-rs. You just have to pull the new code and switch from using `Pin<Box<T>>` to `com::production::ClassAllocation<T>`. This hopefully should fix the issue. Let me know if you run into more issues.

Comment (MarijnS95) [reply macros/support/src/class/class.rs]: > I don't plan on switching back to `ComPtr`. I'll most likely just create a `ClassObject<T>` wrapper object that does the right thing.

Indeed, a wrapper seems the better choice.

> I pushed some new code, and it seems to fix the segfault issue in hassle-rs. You just have to pull the new code and switch from using `Pin<Box<T>>` to `com::production::ClassAllocation<T>`. This hopefully should fix the issue. Let me know if you run into more issues.

Neat, it works now!

I don't really like the extra decrement in `iunknown_impl::to_release_tokens` but see why it's necessary. Could the entire deallocation run through a function on `trait Class`? That makes this case more explicit without having to turn the entire thing in an owned `ClassAllocation` and implicitly dropping it.
Perhaps the `Cell` can be turned into an `Atomic` as well to save get+set trouble? What are concurrency guarantees on these refcounts anyway?

Regarding the `derive(Debug)` question earlier on, perhaps this can be defined again now with:
```rust
impl<T: Class + std::fmt::Debug> std::fmt::Debug for ClassAllocation<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ClassAllocation")
            .field("inner", &self.inner.as_ref())
            .finish()
    }
}
```

Finally, is the `::std::pin::Pin<::std::boxed::Box<Self>>` on the `Deref` and `IUnknown` methods still necessary?

Comment (rylev) [reply macros/support/src/class/class.rs]: I think putting deallocation in the `Class` trait would complicate things. You'd then have to make sure that all fields are still valid to keep around after deallocation. So for instance if your class had a `Vec` as one of its fields, you'd have to wrap it in a `ManuallyDrop` to get things to work. Right now, the class object only actually gets dropped when it should be deallocated. 

We could switch to atomics for the ref count, but I believe more work would be needed to ensure thread safety. Write now we assume that the class is not accessed from different threads. I'm not sure how we'd handle user defined fields that are not thread safe. 

I'll add a debug impl for the `ClassAllocation`.

I'm not sure if `Self: Pin<Box<Self>>` is entirely necessary though it is definitely more correct than `&self` since `Self` in that case is the class not the `ClassAllocation`. I'd prefer to keep it. 

We should make an issue about the `cfg` annotations. I think I'd like to merge this PR before tackling that.

Comment (MarijnS95) [reply macros/support/src/class/class.rs]: > I think putting deallocation in the `Class` trait would complicate things. You'd then have to make sure that all fields are still valid to keep around after deallocation. So for instance if your class had a `Vec` as one of its fields, you'd have to wrap it in a `ManuallyDrop` to get things to work. Right now, the class object only actually gets dropped when it should be deallocated.

I see, `ClassAllocation::drop` and the `release` function can't easily be generalized as the drop takes by mutable reference while `release` intentionally takes the `ClassAllocation` by value to implicitly call `drop()`. No way we should go back to wrapping all members in `ManuallyDrop` and that kind of shenanigans :grin: 

> We could switch to atomics for the ref count, but I believe more work would be needed to ensure thread safety. Write now we assume that the class is not accessed from different threads. I'm not sure how we'd handle user defined fields that are not thread safe.

While tidying up the implementation (getting rid of `get`+`set` pairs) this might give a false sense of thread-safe security, your call.

> I'm not sure if Self: Pin<Box<Self>> is entirely necessary though it is definitely more correct than &self since Self in that case is the class not the ClassAllocation. I'd prefer to keep it.

I wonder if `query`/`query_interface`/`add_ref` should worry about such a thing. Thanks to `Deref` (if it were to return `&Self`) it would anyway be lifetime-bound to a particular `ClassAllocation` instance. Removing it makes the thing segfault, so let's leave it in :shrug:

**Comment by rylev** [comment]
@kennykerr @MarijnS95 I'm going to merge this and add issues for the things we mentioned during the review:
* Casing for COM methods (CamelCase for raw COM methods and snake_case for any pure Rust versions)
* Passing `cfg` done into COM class methods.

**Comment by MarijnS95** [comment]
@rylev Yeah it seems totally ready from my side, thanks for working through the remaining issues :slightly_smiling_face:!

File: macros/support/src/class/class.rs (hunk 7 above)
Comment (MarijnS95) [reply macros/support/src/class/class.rs]: > I'll add a debug impl for the ClassAllocation.

@rylev the idea was to forward the formatting to `inner` in case it supports debugging (as per the code suggestion). Showing a raw pointer to the user seems rather pointless IMO, except for developers on com-rs itself.

The error is pretty helpful too, in case someone forgot to implement debug on their com class but try to `dbg` it:

```rust
error[E0277]: `wrapper::DxcIncludeHandlerWrapper` doesn't implement `std::fmt::Debug`
   --> src/wrapper.rs:270:9
    |
270 |         dbg!(&cls);
    |         ^^^^^^^^^^^ `wrapper::DxcIncludeHandlerWrapper` cannot be formatted using `{:?}`
    |
    = help: the trait `std::fmt::Debug` is not implemented for `wrapper::DxcIncludeHandlerWrapper`
    = note: add `#[derive(Debug)]` or manually implement `std::fmt::Debug`
    = note: required because of the requirements on the impl of `std::fmt::Debug` for `com::production::ClassAllocation<wrapper::DxcIncludeHandlerWrapper>`
```

Unfortunately `+ ?std::fmt::Debug` isn't a thing, otherwise we could have both. For now I'd suggest to implement `Debug` only when `T` implements it too.

Comment (rylev) [reply macros/support/src/class/class.rs]: Thar's fair. The implementation I included was actually very helpful for debugging the segfault issue we were seeing, but I can see how that's potentially not helpful in the general case where COM-rs is not buggy 😉.

Comment (MarijnS95) [reply macros/support/src/class/class.rs]: @rylev I guess a `ClassAllocation@0x<address> { <your struct> }`?

But then we don't want that to segfault before the address has been printed :thinking:

Comment (rylev) [reply macros/support/src/class/class.rs]: Or we keep it as is and add an `AsRef` impl for `ClassAllocation` so it's very easy to get a `&T` which can be debug printed when `T: Debug`. Either way this is now being tracked: https://github.com/microsoft/com-rs/issues/169
